yt_idefix/data_structures.py

```diff
@@ -427,15 +427,173 @@ def _get_time(self):
 
     def _set_code_unit_attributes(self):
         """Conversion between physical units and code units."""
-        length_unit = self.parameters.get("unit_length", 1.0)
-        velocity_unit = self.parameters.get("unit_velocity", 1.0)
-        density_unit = self.parameters.get("unit_density", 1.0)
 
-        self.length_unit = self.quan(length_unit, "cm")
-        self.velocity_unit = self.quan(velocity_unit, "cm/s")
-        density_unit = self.quan(density_unit, "g/cm**3")
+        # User-passed units are preferred to PLUTO's base units for deriving all units
+        pluto_units = {
+            "length_unit": "cm",
+            "velocity_unit": "cm/s",
+            "density_unit": "g/cm**3",
+        }
+        spu = set(pluto_units)
+        uo = self.units_override.copy()
+        if len(uo) < 3:
+            if len(uo) > 0:
+                ytLogger.info(
+                    "Less than 3 units were specified in units_override (got %s). "
+                    "Need to rely on PLUTO's internal units to derive other units",
+                    len(uo),
+                )
+            # Remove those already in units_override
+            spu.difference_update(uo)
+            for unit in spu:
+                uo[unit] = self.quan(self.parameters.get(unit, 1.0), pluto_units[unit])
+                if len(self.units_override) > 0:
+                    try:
+                        self._validate_units_override_keys(uo)
+                        ytLogger.info("Relying on %s: %s.", unit, uo[unit])
+                    except ValueError:
+                        del uo[unit]
+                if len(uo) == 3:
+                    break
+
+        # Then we calculate base units via the units in units_override
+        # The condition statements are essential to prevent getting stuck in infinite recursion
+        # However they are fragile, you must be cautious to change them
+        def _time():
```

Comment:
The ways these functions are designed is not immediately obvious to me. I guess I'd prefer having them named like "_setup_time_unit" ? Also the fact that they have only side-effects (they modify a variable from the outer scope rather than returning values) is a little unsettling. Can you please add some more comments to explain your strategy ?

Reply:
Sure.

Reply:
> Can you please add some more comments to explain your strategy ?

Do you mean the strategy why did I set such condition statements?

I just tried different conditions during the debugging. So I'm sorry to tell you that I don't have a straightforward strategy to constrcut this part. But each one of them is useful if you test it with all the valid combinations.